src/System.Security.Cryptography.Cng/src/System/Security/Cryptography/CngKey.Create.cs

```diff
@@ -0,0 +1,159 @@
+
+                CngUIPolicy uiPolicy = creationParameters.UIPolicy;
+                if (uiPolicy != null)
+                {
+                    InitializeKeyUiPolicyProperties(keyHandle, uiPolicy);
+                }
+
+                // Iterate over the custom properties, setting those as well.
+                foreach (CngProperty property in creationParameters.Parameters)
+                {
+                    byte[] value = property.GetValueWithoutCopying();
+                    int valueLength = (value == null) ? 0 : value.Length;
+                    fixed (byte* pValue = value.MapZeroLengthArrayToNonNullPointer())
+                    {
+                        ErrorCode errorCode = Interop.NCrypt.NCryptSetProperty(keyHandle, property.Name, pValue, valueLength, property.Options);
+                        if (errorCode != ErrorCode.ERROR_SUCCESS)
+                            throw errorCode.ToCryptographicException();
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        ///     Setup the UIPolicy key properties specified in the key creation parameters
+        /// </summary>
+        private static void InitializeKeyUiPolicyProperties(SafeNCryptHandle keyHandle, CngUIPolicy uiPolicy)
+        {
+            unsafe
+            {
+                fixed (char* pinnedCreationTitle = uiPolicy.CreationTitle, 
+                             pinnedFriendlyName = uiPolicy.FriendlyName,
+                             pinnedDescription = uiPolicy.Description)
+                {
+                    NCRYPT_UI_POLICY ncryptUiPolicy = new NCRYPT_UI_POLICY()
+                    {
+                        dwVersion = 1,
+                        dwFlags = uiPolicy.ProtectionLevel,
+                        pszCreationTitle = new IntPtr(pinnedCreationTitle),
+                        pszFriendlyName = new IntPtr(pinnedFriendlyName),
+                        pszDescription = new IntPtr(pinnedDescription),
+                    };
+
+                    ErrorCode errorCode = Interop.NCrypt.NCryptSetProperty(keyHandle, KeyPropertyName.UIPolicy, &ncryptUiPolicy, sizeof(NCRYPT_UI_POLICY), CngPropertyOptions.Persist);
+                    if (errorCode != ErrorCode.ERROR_SUCCESS)
+                        throw errorCode.ToCryptographicException();
+                }
+
+                string useContext = uiPolicy.UseContext;
+                if (useContext != null)
+                {
+                    int useContextByteLength = checked((useContext.Length + 1) * sizeof(char));
+                    fixed (char* pinnedUseContext = useContext)
+                    {
+                        ErrorCode errorCode = Interop.NCrypt.NCryptSetProperty(keyHandle, KeyPropertyName.UseContext, pinnedUseContext, useContextByteLength, CngPropertyOptions.Persist);
+                        if (errorCode != ErrorCode.ERROR_SUCCESS)
+                            throw errorCode.ToCryptographicException();
+                    }
```

Comment:
This is the only place that sets a string key property. That’s why I didn’t add a helper. I’ve found that the helpers in the original code hid a lot of quirky edge-case behavior that then spread virally throughout the code so I didn’t want to go down that path. As it turned out, a single “void*” overload actually worked nicely here compared to the alternatives.